okhttp/src/main/java/okhttp3/internal/connection/AsyncRouteState.java

```diff
@@ -0,0 +1,167 @@
+package okhttp3.internal.connection;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Proxy;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import okhttp3.Address;
+import okhttp3.AsyncDns;
+import okhttp3.Call;
+import okhttp3.EventListener;
+import okhttp3.Route;
+
+public class AsyncRouteState implements RouteState {
+
+  private final Address address;
+  private final AsyncDns asyncDns;
+
+  private CompletableFuture<List<InetSocketAddress>> futureInetSocketAddresses;
+
+  private CompletableFuture<AllRoutes> futureComputedRoutes;
+
+  private AllRoutes computedRoutes;
+
+  public AsyncRouteState(Address address) {
+    this.address = address;
+    this.asyncDns = (AsyncDns) address.dns();
+  }
+
+  @Override public boolean hasPostponedRoutes() {
+    if (futureComputedRoutes.isDone()) {
+      AllRoutes allRoutes = launderedGetComputedRoutes();
+      return !allRoutes.postponedRoutes.isEmpty();
+    }
+
+    return false;
+  }
+
+  private AllRoutes launderedGetComputedRoutes() {
+    try {
+      return getComputedRoutes();
+    } catch (IOException e) {
+      if (e instanceof UnknownHostException)
+        throw new UnknownHostRuntimeException(e);
+
+      throw new RuntimeException(e);
+    }
+  }
+
+  private AllRoutes getComputedRoutes() throws IOException {
+    try {
+      if (computedRoutes == null)
+        computedRoutes = futureComputedRoutes.get();
```

Comment:
This is my main concern, a blocking call here is making this look simple while it's actually problematic.  e.g. if we use a single executor for both the OkHttp client and the Dns implementation, its simpler for users.  But now we have something capable of deadlocks IIUC.